sdk/resource/resource.go

```diff
@@ -51,29 +74,71 @@ func Merge(a, b *Resource) *Resource {
 	if b == nil {
 		return a
 	}
-	res := &Resource{
-		labels: map[core.Key]core.Value{},
+
+	n := len(a.sorted)
+	if len(b.sorted) > len(a.sorted) {
+		n = len(b.sorted)
+	}
+	// At a minimum the merge will be as large as the largest resource.
+	s := make([]core.KeyValue, 0, n)
+	k := make(map[core.Key]struct{}, n)
+	ai, bi := 0, 0
+	for ; ai < len(a.sorted) && bi < len(b.sorted); ai, bi = ai+1, bi+1 {
+		akv := a.sorted[ai]
+		s = append(s, akv)
+		k[akv.Key] = struct{}{}
+
+		bkv := b.sorted[bi]
+		if _, ok := a.keys[bkv.Key]; ok {
+			// a overwrites b.
+			continue
+		}
+		k[bkv.Key] = struct{}{}
+		s = append(s, bkv)
 	}
-	for k, v := range b.labels {
-		res.labels[k] = v
+
+	for ; ai < len(a.sorted); ai++ {
+		akv := a.sorted[ai]
+		s = append(s, akv)
+		k[akv.Key] = struct{}{}
 	}
-	// labels from resource a overwrite labels from resource b.
-	for k, v := range a.labels {
-		res.labels[k] = v
+
+	for ; bi < len(b.sorted); bi++ {
+		bkv := b.sorted[bi]
+		if _, ok := k[bkv.Key]; ok {
+			continue
+		}
+		k[bkv.Key] = struct{}{}
+		s = append(s, bkv)
 	}
-	return res
-}
 
-// Attributes returns a copy of attributes from the resource.
-func (r Resource) Attributes() []core.KeyValue {
-	attrs := make([]core.KeyValue, 0, len(r.labels))
-	for k, v := range r.labels {
-		attrs = append(attrs, core.KeyValue{Key: k, Value: v})
+	sort.Slice(s, func(i, j int) bool { return s[i].Key < s[j].Key })
+	res := &Resource{
+		keys:   k,
+		sorted: s,
 	}
-	return attrs
+	res.str = buildResourceString(res.sorted)
+
+	return res
 }
 
-// Equal returns true if other Resource is the equal to r.
-func (r Resource) Equal(other Resource) bool {
-	return reflect.DeepEqual(r.labels, other.labels)
+// buildResourceString returns a string representation of a Resource
+// containing kvs.
+func buildResourceString(kvs []core.KeyValue) string {
```

Comment:
The purpose of this function is to return a unique string for the passed key values. This needs to do some escaping of the special characters (`=` and `,` and the escape char [likely `\`]). Otherwise the following slices return the same string (`Resource(foo=bar,baz=quux)`):

```go
[]core.KeyValue{key.String("foo", "bar"), key.String("baz", "quux")}
[]core.KeyValue{key.String("foo", "bar,baz=quux")}
[]core.KeyValue{key.String("foo=bar,baz", "quux")}
```

You could integrate those in the `TestString` test.

Reply:
Good point, will do.